infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -0,0 +1,60 @@
+package org.corfudb.infrastructure.logreplication.infrastructure;
+
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.runtime.collections.TxnContext;
+
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once upgrades are done it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Practically this means any code that is running Active/Standby logic is considered "V1"
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
+ *                  ┌────────────────────────────────┐
+ *                  │ isClusterAtV2 = false (in-mem) │
+ *                  └──────────────────┬─────────────┘
+ *                                     │
+ *                     ┌───────────────▼───────────┐
+ *                   ┌─┤                           ├─┐
+ *         Yes ┌─────┴─┴┐  isRollingUpgradeOn() ? ┌┴─┴──────┐ No
+ *             │        └─────────────────────────┘         │
+ *             │                                            │
+ * ┌───────────▼───────────┐                         ┌──────▼───────────────┐
+ * │ isClusterAtV2 = false │                         │ isClusterAtV2 = true │
+ * └───────────┬───────────┘                         └───────┬──────────────┘
+ *             │            ┌──────────────────────┐         │
+ *             └───────────►│ start Log Replicator │◄────────┘
+ *                          └──────────────────────┘
+ */
+public class LRRollingUpgradeHandler {
+
+    private volatile boolean isClusterAtV2 = false;
+    ILogReplicationVersionAdapter versionAdapter;
+    public LRRollingUpgradeHandler(ILogReplicationVersionAdapter versionAdapter) {
+        this.versionAdapter = versionAdapter;
+    }
+
+    public boolean isLRUpgradeInProgress(TxnContext txnContext) {
+        if (isClusterAtV2) {
+            return false;
+        }
```

Comment:
So it looks like here `isClusterAtV2` indicates not only if the rolling upgrade is complete or not, it also indicates the data migration is done, and LR in this cluster could start to operate completely in V2.

So maybe we can rename it to be something more self-explanatory like `isLRReadyForV2` or `isClusterReadyForV2`? Since `isClusterAtV2` gives me an impression that it has pretty similar semantics as `isClusterUpgradeInProgress`

Reply:
Yes `isClusterAtV2` indicates both that rolling upgrade has completed & data migration is complete OR it could mean that the entire cluster is a new instantiation at V2 with no necessity for upgrade.

Sure I will rename it to `isLRReadyForV2` if that seems more self-explanatory. thanks